ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/KnownPeers.hs

```diff
@@ -135,16 +142,24 @@ belowTarget actions
 
 
 jobPeerShare :: forall m peeraddr peerconn.
-                (MonadAsync m, MonadTimer m, Ord peeraddr)
+                (MonadAsync m, MonadTimer m, Ord peeraddr, Hashable peeraddr)
              => PeerSelectionActions peeraddr peerconn m
              -> PeerSelectionPolicy peeraddr m
+             -> Int
+             -> Int
              -> PeerSharingAmount
              -> [peeraddr]
              -> Job () m (Completion m peeraddr peerconn)
 jobPeerShare PeerSelectionActions{requestPeerShare}
-             PeerSelectionPolicy{..} =
+             PeerSelectionPolicy{..} salt maxAmount =
     \amount peers -> Job (jobPhase1 amount peers) (handler peers) () "peerSharePhase1"
   where
+    -- Return n random peers from a list of peers
+    takeNPeers :: Int -> [peeraddr] -> [peeraddr]
+    takeNPeers n addrs = take n $
+      sortBy (\a b -> compare (hashWithSalt salt a) (hashWithSalt salt b))
+      addrs
```

Comment:
With this, for similar lists we will pick more or less the same peers, is that right?

Reply:
Yes this will give stable results, but importantly this will not allow a peer to make it more likely to be included by using a small IP address.

Reply:
Every peershare run will use a new salt so if we are presented with similar lists of usable peers we should end up with a different order. Also after the first time we see the list of peers some will be added so they will no longer be part of the list of usable peers to choose from.

Reply:
Thanks!

Reply:
We are also removing peers that are known to us (or are big ledger peers), so each time we will be adding new ones. This also makes sure that each time we pick only useful ones.  But this makes it less likely that we will pick the same peer that we already tried and decided it wasn't good enough (it was dropped by churn).

Reply:
@karknu this is subtle enough to deserve an explanation in the code.

Reply:
I've expanded on the comment for the function.